hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -1387,45 +1387,82 @@ def should_replace_byte(b):
+
+        Commands are:
+
+            * "-", subtract one from this block.
+            * "X", delete this block
+
+        If a command does not apply (currently only because it's - on a zero
+        block) the block will be silently skipped over.
+
+        This method runs the block program in ``description`` at block index
+        ``i`` on the ConjectureData ``original``. If ``repeats > 1`` then it
+        will attempt to approximate the results of running it that many times.
+        """
+        if i + len(description) > len(original.blocks) or i < 0:
+            return False
+        attempt = bytearray(original.buffer)
+        for _ in hrange(repeats):
+            for k, d in reversed(list(enumerate(description))):
+                j = i + k
+                u, v = original.blocks[j].bounds
+                if v > len(attempt):
+                    return False
+                if d == "-":
+                    value = int_from_bytes(attempt[u:v])
+                    if value == 0:
+                        return False
+                    else:
+                        attempt[u:v] = int_to_bytes(value - 1, v - u)
+                elif d == "X":
+                    del attempt[u:v]
+                else:  # pragma: no cover
+                    assert False, "Unrecognised command %r" % (d,)
+        return self.incorporate_new_buffer(attempt)
+
 
 def block_program(description):
-    """Mini-DSL for block rewriting. A sequence of commands that will be run
-    over all contiguous sequences of blocks of the description length in order.
-    Commands are:
-
-        * ".", keep this block unchanged
-        * "-", subtract one from this block.
-        * "0", replace this block with zero
-        * "X", delete this block
-
-    If a command does not apply (currently only because it's - on a zero
-    block) the block will be silently skipped over. As a side effect of
-    running a block program its score will be updated.
-    """
+    """Defines a shrink pass that runs the block program ``description``
+    at every block index."""
+    n = len(description)
 
     def generate_arguments(self):
-        return [(i,) for i in hrange(len(self.blocks))]
+        return [(i,) for i in hrange(len(self.blocks) - n)]
 
     def run(self, i):
-        n = len(description)
-        if i + n > len(self.shrink_target.blocks):
+        """Adaptively attempt to run the block program at the current
+        index. If this successfully applies the block program ``k`` times
+        then this runs in ``O(log(k))`` test function calls."""
+        assert i + n < len(self.shrink_target.blocks)
+        # Because we run in a random order we will often find ourselves in the middle
+        # of a region where we could run the block program. We thus start by moving
+        # left to the beginning of that region if possible in order to to start from
+        # the beginning of that region..
+        move_left = find_integer(
```

Comment:
The fact that `move_left` is offset by 1 is pretty confusing.

It would be better to arrange for `move_left` to be 0 when it succeeds at `i`, or -1 if it fails. (Then adjust the rest of the function accordingly.)

This could be as simple as `(-1) + find_integer`, or potentially a wrapper around `find_integer` that does the offsetting internally.

Reply:
I realised an easy resolution: We can just not do the offset by 1 - if it's zero then we just don't change `i` here and `i` will get run at the next stage when we do `find_integer` on the number of repetitions.